### ISSUE TITLE
Stack overflow on saving node graph with certain target parameterization

### ISSUE THREAD

**Issue by lucianm** [fastbuild/fastbuild#415]
Hello Franta,

the changes done somewhere between version 0.94 and 0.95 had fbuild crashing on a complex project structure of ours, which used to work with 0.94. At first I was left with no other mean of investigation than running fbuild through the debugger and I found out that the effect of crashing was caused by a stack overflow in the mutual, reccuring calls to the two overloads _NodeGraph::SaveRecurse_. Fortunately, that gave me the possibility to observe that every time a few targets which I had placed at the very bottom of the dependency tree were involved. They are **Exec** functions in which in former versions of FASTBuild I needed to provide both **ExecInput** and **ExecOutput** pointing to the same folder / file. Of course, one can argue that this should not be necessary (well, in the end it wasn't, I could actually just remove the optional _ExecInput_ from them and the scripts called in _ExecExecutable_ worked well, too), but nevertheless, I think that this should have rather been a BFF parsing error reported to the user, maybe it can be considered a syntactical error in the BFF having _ExecInput_ and _ExecOuput_ pointing to the exact same string, or maybe it is truly an error in walking the whole dependency graph when saving the database and only not encountering this situation does not expose it. You surely know best where to look after this :-)

Greetings,
Lucian

**Comment by ffulin** [comment]
Thanks for the bug report. Of course FASTBuild should never crash, and it does sound like an error should be emitted in this case. I will investigate.

**Comment by ffulin** [comment]
I investigated and reproduced this problem. If any function was invoked with properties such that any Input was the same as the Output, then the node would end up depending on itself and a stack overflow would occur during saving of the dependency DB.

I've added code to detect this in a general way (https://github.com/fastbuild/fastbuild/commit/3567c53dfa6734fdad7fb3e52ec08709f53b7a41) and added a regression test.

This fix will be in the next release (v0.96). I'll leave this issue open until that version is released. Thanks for reporting the bug!

**Comment by ffulin** [comment]
When a cyclic dependency is detected, an error is now emitted during parsing, and parsing halts:
```C++
c:\Test\fbuild.bff(5,1): FASTBuild Error #1043 - Cyclic dependency detected for node 'C:\Test\file.txt'.
Exec( 'bad' )
^
\--here
```

**Comment by lucianm** [comment]
Great, I'm glad it will be fixed in the next version!

**Comment by ffulin** [comment]
A version containing a fix for this issue has now been released (v0.96).
